scripts/offline_data_processing.py

```diff
@@ -0,0 +1,264 @@
+# Standard
+import logging
+import os
+import sys
+import traceback
+
+# Third Party
+from transformers import (
+    AutoTokenizer,
+    GPT2Tokenizer,
+    GPTNeoXTokenizerFast,
+    LlamaTokenizer,
+    LlamaTokenizerFast,
+)
+
+# Local
+from tuning.config import configs
+from tuning.data.setup_dataprocessor import process_dataargs
+from tuning.sft_trainer import get_parser
+from tuning.utils.error_logging import USER_ERROR_EXIT_CODE, write_termination_log
+from tuning.utils.logging import set_log_level
+
+
+def save_dataset_shards(
+    dataset, output_dir: str, num_shards: int, dataset_name: str
+) -> None:
+    """
+    Saves the given dataset in the specified number of shards.
+
+    Args:
+        dataset: The dataset to shard and save.
+        output_dir (str): Directory to save the dataset shards.
+        num_shards (int): Number of shards to create.
+        dataset_name (str): Name of the dataset (used for logging).
+    """
+    os.makedirs(output_dir, exist_ok=True)
+    for shard_idx in range(num_shards):
+        shard = dataset.shard(index=shard_idx, num_shards=num_shards)
+        shard_path = os.path.join(output_dir, f"ds_{shard_idx:05d}.parquet")
+        shard.to_parquet(shard_path)
+    logging.info("Dumped %d shards of %s at %s", num_shards, dataset_name, output_dir)
+
+
+def get_processed_dataset(
+    model_args: configs.ModelArguments,
+    data_args: configs.DataArguments,
+    train_args: configs.TrainingArguments,
+):
+    """
+    Processes the dataset based on data config yaml.
+
+    Args:
+        model_args (configs.ModelArguments): Model configuration arguments.
+        data_args (configs.DataArguments): Data configuration arguments.
+        train_args (configs.TrainingArguments): Training configuration arguments.
+
+    Returns:
+        tuple: A tuple containing the formatted training dataset and validation dataset.
+    """
+    # Set log level for this function
+    train_args, logger = set_log_level(train_args, "get_processed_dataset")
+
+    logger.info(
+        "Starting dataset processing with model_args: %s, data_args: %s, training_args: %s",
+        model_args,
+        data_args,
+        train_args,
+    )
+
+    # Load tokenizer for the model
+    tokenizer_path = model_args.tokenizer_name_or_path or model_args.model_name_or_path
+    logger.debug("Loading tokenizer from %s", tokenizer_path)
+    tokenizer = AutoTokenizer.from_pretrained(
+        tokenizer_path,
+        cache_dir=train_args.cache_dir,
+        use_fast=True,
+        legacy=True,
+    )
+    logger.debug("Tokenizer loaded successfully.")
+
+    # Add chat_template to the tokenizer if provided
+    if data_args.chat_template:
```

Comment:
For now this all is good but we should see how to create some function in fms-hf-tuning code to not execute all these things twice...easier to maintain for us if this script is minimal.